storage/backend: support MAP_POPULATE for unix

gyuho opened the pull request:
/cc @xiang90 

This adds build constraints in order to pass memory-map flags to bolt.Option.
If backend package passes syscall.MAP_POPULATE flag, the boltdb does read-ahead
which speeds up entire-database read, which then leads to faster storage
Restore. Benchmark result shows for 4GB, it opens and loads 6x faster in SSD,
and 12x faster in HDD. For 2GB, 1.6x faster with MAP_POPULATE in SSD.


<br>
<hr>
##### Background

storage package should do read ahead to make kernal preload some pages into
page cache and speed up the recover reads in etcd.

Related issue:
- https://github.com/coreos/etcd/issues/3786

<br>
##### Test script

```
	start := time.Now()
	kv := storage.New("test.db", &fg{})
	kv.Restore()
	fmt.Println("kv.Restore took:", time.Since(start))
```


<br>
##### Results: SSD (local machine)

```
Ubuntu 15.10 with SSD
model name  : Intel(R) Core(TM) i7-4910MQ CPU @ 2.90GHz
go version go1.5.1 linux/amd64
linux kernel version: 4.2.0-18-generic

open + load(restore) 4GB data of 1 million random bytes keys:
WITHOUT MAP_POPULATE: 2 min 5 sec (open 3.706302ms, restore 1m58.015088268s)
WITH    MAP_POPULATE: 21 sec (open 7.771817481s, restore 13.6191382s)

open + load(restore) 2GB data of 1 million simple integer keys:
WITHOUT MAP_POPULATE: 18 sec (open 10.893178ms, restore 17.197815402s)
WITH    MAP_POPULATE: 11 sec (open 3.923564768s, restore 6.810750348s)
```


<br>
##### Results: HDD (GCE)

```
Debian GNU/Linux 7.9 (wheezy) with HDD
machine type: n1-highmem-2 (2 vCPUs, 13 GB memory) / Intel Haswell
model name      : Intel(R) Xeon(R) CPU @ 2.30GHz
go version go1.5.1 linux/amd64
linux kernel version: 3.16.0-0.bpo.4-amd64

open + load(restore) 4GB data of 1 million random bytes keys:
WITHOUT MAP_POPULATE: 9 min 10 sec (open 28.758918ms, restore 8m40.518096833s)
WITH    MAP_POPULATE: 46 sec (open 36.944924626s, restore 9.460929924s)
```

xiang90 commented:
#2698 is not related.

gyuho commented:
@xiang90 Just removed that mention. Thanks!

gyuho commented:
And I plan to buy a Window machine (I am using ubuntu machine now). Once I get it, I can try to support Windows as well.

xiang90 commented on storage/backend/boltoption_darwin.go:
```diff
@@ -0,0 +1,19 @@
+
+import "github.com/coreos/etcd/Godeps/_workspace/src/github.com/boltdb/bolt"
+
+var boltOption *bolt.Options = nil
```
let's be more specific? boltOpenOptions?

xiang90 commented on storage/backend/boltoption_solaris.go:
```diff
@@ -0,0 +1,25 @@
+	"github.com/coreos/etcd/Godeps/_workspace/src/github.com/boltdb/bolt"
+)
+
+var boltOption = &bolt.Options{MmapFlags: syscall.MAP_POPULATE}
```
can we add comments about why do we want to enable MAP_POPULATE?

xiang90 commented:
@gyuho Thanks for the extensive analysis. The code looks good to me.

gyuho commented:
I added comments as you suggested. Please take a look again.

Thanks @xiang90

yichengq commented on storage/backend/boltoption_unix.go:
```diff
@@ -0,0 +1,28 @@
+// syscall.MAP_POPULATE on Linux 2.6.23+ does sequential read-ahead
+// which can speed up entire-database read with boltdb. We want to
+// enable MAP_POPULATE for faster storage key-value store restore.
+var boltOptions = &bolt.Options{MmapFlags: syscall.MAP_POPULATE}
```
For curiosity, why don't merge the file _unix and _solaris into one? They are have the same content.

gyuho replied:
You are right. I combined two into one `_unix.go`. Thanks!

xiang90 commented on storage/backend/boltoption_unix.go:
```diff
@@ -0,0 +1,28 @@
+	"github.com/coreos/etcd/Godeps/_workspace/src/github.com/boltdb/bolt"
+)
+
+// syscall.MAP_POPULATE on Linux 2.6.23+ does sequential read-ahead
```
what happens if people with linux kernel < 2.6.23?

xiang90 commented on storage/backend/boltoption_darwin.go:
```diff
@@ -0,0 +1,19 @@
+
+import "github.com/coreos/etcd/Godeps/_workspace/src/github.com/boltdb/bolt"
+
+var boltOptions *bolt.Options = nil
```
sorry. i should be more specific. I mean boltOpenOptions. `Open` means a lot in the case. Boltdb has other options too.

gyuho replied:
got it

xiang90 commented on storage/backend/boltoption_solaris.go:
```diff
@@ -0,0 +1,28 @@
+	"github.com/coreos/etcd/Godeps/_workspace/src/github.com/boltdb/bolt"
+)
+
+// syscall.MAP_POPULATE on Linux 2.6.23+ does sequential read-ahead
```
This does not make any sense in Solaris doc at all, right?

gyuho replied:
Right, I will remove it.

xiang90 commented on storage/backend/boltoption_solaris.go:
```diff
@@ -0,0 +1,28 @@
+// syscall.MAP_POPULATE on Linux 2.6.23+ does sequential read-ahead
+// which can speed up entire-database read with boltdb. We want to
+// enable MAP_POPULATE for faster storage key-value store restore.
+var boltOptions = &bolt.Options{MmapFlags: syscall.MAP_POPULATE}
```
and why do we need another file for solaris besides unix?

gyuho replied:
That's how boltdb does. And since I specify the OS by file name suffix, we need to list both unix and solaris, I think. Do you have a better way? Actually, the file name suffix overrides build tags.

xiang90 replied:
bolt does this just because bolt has solairs specific logic. Please check their flock implementation. We do not have here.

gyuho replied:
Ok I will.

gyuho commented on storage/backend/boltoption_unix.go:
```diff
@@ -0,0 +1,30 @@
+// which can speed up entire-database read with boltdb. We want to
+// enable MAP_POPULATE for faster key-value store recovery in storage
+// package. If your kernel is lower than 2.6.23, mmap might silently
+// ignored this flag. Please update your kernel to prevent this.
```
Kernel more than 8-year old (https://github.com/torvalds/linux/releases/tag/v2.6.23) might be affected :(

gyuho commented:
@xiang90 

> (example: source_windows_amd64.go) where GOOS and GOARCH represent any known operating system and architecture values respectively, then the file is considered to have an implicit build constraint requiring those terms (in addition to any explicit constraints in the file). https://golang.org/pkg/go/build/#ArchChar

Explains if OS is specified by filename suffix, it will be included and the OSes with build tags `build +` are also added when building. So `boltoption_unix.go` would be included when compiled in unix. I added `linux` and `solaris` because unix is for some reason not included here https://github.com/golang/go/blob/master/src/go/build/syslist.go#L7 which is used to check the OS is good or not. But looking at Go standard libraries, it should be ok. I added it just to be safe. (https://github.com/golang/go/blob/f78a4c84ac8ed44aaf331989aa32e40081fd8f13/src/net/cgo_unix.go, https://github.com/golang/go/tree/73f329f47236c6952235db395802810156a7bd65/src/syscall)

xiang90 commented on storage/backend/boltoption_unix.go:
```diff
@@ -0,0 +1,31 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build linux solaris
```
How about freebsd?

gyuho replied:
https://www.freebsd.org/cgi/man.cgi?query=mmap&apropos=0&sektion=2&manpath=FreeBSD+11-current&arch=default&format=html

FreeBSD kernel doesn't have MAP_POPULATE flag, as far as I know. Should we add a separate build tag with `nil` for freebsd?

xiang90 replied:
probably we should. just like the darwin one.

gyuho replied:
Got it.

gyuho commented on storage/backend/boltoption_freebsd.go:
```diff
@@ -0,0 +1,19 @@
+
+import "github.com/coreos/etcd/Godeps/_workspace/src/github.com/boltdb/bolt"
+
+var boltOpenOptions *bolt.Options = nil
```
Just added. Thanks Xiang

xiang90 commented:
LGTM. Will merge when tests pass.

gyuho commented:
Awesome. Thanks,
